### PR TITLE
OHRM5X-1529: Develop maintenance - purge candidate API

### PR DESCRIPTION
**Pull request by devishke-orange**
PR contains:
- OHRM5X-1529: BE - Purge Candidates

P1 fixes:
- OHRM5X-1531: [Broken][Tracker Logs] Tracker logs added are not getting displayed.
- OHRM5X-1533: [Broken][Time - Project info] Project info list view page keeps loading when project admins are added.

### CODE REVIEW

File: src/plugins/orangehrmMaintenancePlugin/Service/PurgeEmployeeService.php [hunk 1]
```diff
@@ -75,6 +76,34 @@ public function purgeEmployeeData(int $empNumber): void
         }
     }
 
+    /**
+     * @param int $vacancyNumber
+     * @throws TransactionException
+     */
+    public function purgeCandidateData(int $vacancyNumber): void
```
Comment (devishke-orange) [inline src/plugins/orangehrmMaintenancePlugin/Service/PurgeEmployeeService.php:83]: this function is a duplicate of `purgeEmployeeData` function. only different is the file name `self::GDPR_PURGE_CANDIDATE`. This is how it was done in 4.x. Shall I combine the two functions and have an argument for the file name?

File: src/plugins/orangehrmMaintenancePlugin/PurgeStrategy/DestroyCandidatePurgeStrategy.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    public function purge(int $vacancyNumber): void
+    {
+        $matchByValues = $this->getMatchByValues($vacancyNumber);
+        $purgeEntities = $this->getEntityRecords($matchByValues, $this->getEntityClassName());
+        foreach ($purgeEntities as $purgeEntity) {
+            /** @var Candidate $purgeEntity */
+            if (!$purgeEntity->isConsentToKeepData()) {
+                $this->getEntityManager()->remove($purgeEntity);
+            }
+        }
```
Comment (RajithaKumara) [inline src/plugins/orangehrmMaintenancePlugin/PurgeStrategy/DestroyCandidatePurgeStrategy.php:38]: 1. ref: https://www.doctrine-project.org/projects/doctrine-orm/en/2.11/reference/batch-processing.html#bulk-deletes
2. or we can go for query builder,

```php
        $qb = $this->createQueryBuilder(Candidate::class, 'candidate');
        $qb->leftJoin('candidate.candidateVacancy', 'candidateVacancy');
        $qb->delete()
            ->andWhere('candidate.consentToKeepData = :consentToKeepData')
            ->setParameter('consentToKeepData', false)
            ->andWhere('candidateVacancy.vacancy = :vacancyId')
            ->setParameter('vacancyId', $vacancyId);
```
not sure about the 2nd approach with the design

File: src/plugins/orangehrmMaintenancePlugin/config/gdpr_purge_candidate_strategy.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+Entities :
+  Candidate:
```
Comment (RajithaKumara) [inline src/plugins/orangehrmMaintenancePlugin/config/gdpr_purge_candidate_strategy.yaml:2]: How about the candidate attachment and history. Does cascade remove?

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: installer/Migration/V5_1_0/permission/api.yaml [hunk 4]
```diff
@@ -489,3 +489,15 @@ apiv2_corporate_branding_theme_preview:
     delete: false
   permissions:
     - { role: Admin, permission: { read: false, create: true, update: false, delete: false } }
+
```
[the remark this answers is not in the document]
Comment (devishke-orange) [reply installer/Migration/V5_1_0/permission/api.yaml:492]: added

File: src/plugins/orangehrmMaintenancePlugin/Api/PurgeCandidateAPI.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+use OrangeHRM\Core\Api\V2\Validator\ParamRuleCollection;
+use OrangeHRM\Core\Api\V2\Validator\Rule;
+use OrangeHRM\Core\Api\V2\Validator\Rules;
+use OrangeHRM\Maintenance\Api\Model\PurgeEmployeeModel;
+use OrangeHRM\Maintenance\Service\PurgeEmployeeService;
+use OrangeHRM\ORM\Exception\TransactionException;
+
+class PurgeCandidateAPI extends Endpoint implements CollectionEndpoint
+{
+    private ?PurgeEmployeeService $purgeEmployeeService = null;
```
[the remark this answers is not in the document]
Comment (devishke-orange) [reply src/plugins/orangehrmMaintenancePlugin/Api/PurgeCandidateAPI.php]: renamed to `PurgeService`

File: installer/Migration/V5_1_0/permission/api.yaml [hunk 6]
```diff
@@ -489,3 +489,15 @@ apiv2_corporate_branding_theme_preview:
...
+apiv2_maintenance_purge_candidate:
+  description: 'API-v2 Maintenance - Purge Candidate'
+  api: OrangeHRM\Maintenance\Api\PurgeCandidateAPI
+  module: maintenance
+  allowed:
+    read: true
```
[the remark this answers is not in the document]
Comment (devishke-orange) [reply installer/Migration/V5_1_0/permission/api.yaml:502]: yes same as purge employee. not sure why I put it :sweat_smile:

File: src/plugins/orangehrmMaintenancePlugin/PurgeStrategy/DestroyCandidatePurgeStrategy.php (hunk 2 above)
Comment (devishke-orange) [reply src/plugins/orangehrmMaintenancePlugin/PurgeStrategy/DestroyCandidatePurgeStrategy.php]: Is what I've done okay? @RajithaKumara

File: src/plugins/orangehrmMaintenancePlugin/config/gdpr_purge_candidate_strategy.yaml (hunk 3 above)
Comment (devishke-orange) [reply src/plugins/orangehrmMaintenancePlugin/config/gdpr_purge_candidate_strategy.yaml:2]: Yes. Test cases are added to service test

File: src/plugins/orangehrmMaintenancePlugin/Api/PurgeCandidateAPI.php [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    /**
+     * @inheritDoc
+     */
+    public function getValidationRuleForDelete(): ParamRuleCollection
+    {
+        return new ParamRuleCollection(
+            new ParamRule(
+                self::PARAMETER_VACANCY_ID,
+                new Rule(Rules::POSITIVE),
+                new Rule(Rules::ENTITY_ID_EXISTS, [Vacancy::class])
```
Comment (RajithaKumara) [inline src/plugins/orangehrmMaintenancePlugin/Api/PurgeCandidateAPI.php:76]: shall we remove this and handle 404 in `delete` method :thinking:

File: src/plugins/orangehrmMaintenancePlugin/PurgeStrategy/DestroyCandidatePurgeStrategy.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        $purgeIds = [];
+        foreach ($purgeEntities as $purgeEntity) {
+            /** @var Candidate $purgeEntity */
+            if (!$purgeEntity->isConsentToKeepData()) {
+                $purgeIds[] = $purgeEntity->getId();
+            }
+        }
+        if (!empty($purgeIds)) {
+            $candidateClass = Candidate::class;
+            $this->getEntityManager()->createQuery("delete from $candidateClass c where c.id in (:ids)")
```
Comment (RajithaKumara) [inline src/plugins/orangehrmMaintenancePlugin/PurgeStrategy/DestroyCandidatePurgeStrategy.php:42]: Oops. shall we use query builder
```php
        $qb = $this->createQueryBuilder(Candidate::class, 'candidate');
        $qb->leftJoin('candidate.candidateVacancy', 'candidateVacancy');
        $qb->delete()
            ->andWhere('candidate.consentToKeepData = :consentToKeepData')
            ->setParameter('consentToKeepData', false)
            ->andWhere('candidateVacancy.vacancy = :vacancyId')
            ->setParameter('vacancyId', $vacancyId);
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**